maven-caching/src/main/java/org/apache/maven/caching/DefaultProjectInputCalculator.java

```diff
@@ -0,0 +1,127 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import org.apache.maven.SessionScoped;
+import org.apache.maven.caching.checksum.MavenProjectInput;
+import org.apache.maven.caching.xml.CacheConfig;
+import org.apache.maven.caching.xml.build.ProjectsInputInfo;
+import org.apache.maven.execution.MavenSession;
+import org.apache.maven.lifecycle.internal.builder.BuilderCommon;
+import org.apache.maven.project.MavenProject;
+import org.apache.maven.repository.RepositorySystem;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import java.util.LinkedHashSet;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+@SessionScoped
+@Named
+public class DefaultProjectInputCalculator implements ProjectInputCalculator
+{
+    private static final Logger LOGGER = LoggerFactory.getLogger( DefaultProjectInputCalculator.class );
+
+    private final MavenSession mavenSession;
+    private final RemoteCacheRepository remoteCache;
+    private final CacheConfig cacheConfig;
+    private final RepositorySystem repoSystem;
+    private final NormalizedModelProvider normalizedModelProvider;
+    private final MultiModuleSupport multiModuleSupport;
+
+    private final ConcurrentMap<String, ProjectsInputInfo> checkSumMap = new ConcurrentHashMap<>();
+
+    private static final ThreadLocal<Set<String>> CURRENTLY_CALCULATING = ThreadLocal.withInitial(
+            LinkedHashSet::new
+    );
+
+    @Inject
+    public DefaultProjectInputCalculator( MavenSession mavenSession,
+                                          RemoteCacheRepository remoteCache,
+                                          CacheConfig cacheConfig,
+                                          RepositorySystem repoSystem,
+                                          NormalizedModelProvider rawModelProvider,
+                                          MultiModuleSupport multiModuleSupport )
+    {
+        this.mavenSession = mavenSession;
+        this.remoteCache = remoteCache;
+        this.cacheConfig = cacheConfig;
+        this.repoSystem = repoSystem;
+        this.normalizedModelProvider = rawModelProvider;
+        this.multiModuleSupport = multiModuleSupport;
+    }
+
+    @Override
+    public ProjectsInputInfo calculateInput( MavenProject project )
+    {
+        LOGGER.info( "Going to calculate checksum for project [groupId=" + project.getGroupId()
+                + ", artifactId=" + project.getArtifactId() + "]" );
+
+        String key = BuilderCommon.getKey( project );
+        ProjectsInputInfo projectsInputInfo = checkSumMap.get( key );
```

Comment:
looks exactly like computeIfAbsent

Reply:
> could be simplified with computeIfAbsent

recursive computeIfAbsent calls are not allowed in CHM, this could lead to IllegalStateException("Recursive update")

here is the example to reproduce

```
    private static ConcurrentMap<Integer, String> suspect = new ConcurrentHashMap<>();

    private static String recursiveUpdate( int i )
    {
        if ( i == 16 )
        {
            return String.valueOf( i );
        }
        return suspect.computeIfAbsent( i, k -> recursiveUpdate( k++ ) );
    }


    public static void main(String[] args) {
        recursiveUpdate( 0 );

        System.out.println( "Never printed" );
    }
```
In jdk8 there could be infinite loop
https://stackoverflow.com/questions/28840047/recursive-concurrenthashmap-computeifabsent-call-never-terminates-bug-or-fea

Reply:
Interesting, thank you. worth to leave a code comment because it is not obvious (and recursive calls too tbh) - probably i'm not the last one of thinking to refactor this.

Reply:
> worth to leave a code comment because it is not obvious (and recursive calls not immediately visible too tbh) - probably i'm not the last one of thinking to refactor this.

added comment